### PR TITLE
8341880: RISC-V: riscv_vector.h native build fails with gcc13 after JDK-8320500

### PR DESCRIPTION
**Pull request by sendaoYan**
Hi all,
The file `src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c` introduced by [JDK-8341880](https://bugs.openjdk.org/browse/JDK-8341880) native build fails by fedora OS shipped gcc13.
Gcc13 doesn't have `__riscv_v_intrinsic` macro by default, and do have `__riscv_v_intrinsic` macro with option: `-march=rv64gcv`. libsleef only supported from llvm version 17 and gcc version 14 onwards, so we need to check the compiler version before use the `__riscv_v_intrinsic`.

image

gcc version infomation:
```
/usr/bin/gcc -v
Using built-in specs.
COLLECT_GCC=/usr/bin/gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/riscv64-redhat-linux/13/lto-wrapper
Target: riscv64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,m2,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --enable-libstdcxx-backtrace --with-libstdcxx-zoneinfo=/usr/share/zoneinfo --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-13.2.1-20230728/obj-riscv64-redhat-linux/isl-install --enable-gnu-indirect-function --with-arch=rv64gc --with-abi=lp64d --with-multilib-list=lp64d --build=riscv64-redhat-linux --with-build-config=bootstrap-lto --enable-link-serialization=1
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.2.1 20230728 (Red Hat 13.2.1-1) (GCC)
```

Additonal testing:

- [x] riscv64 native release debug-level build with gcc13
- [x] riscv64 native fastdebug debug-level build with gcc13
- [x] riscv64 native release debug-level build with gcc14
- [ ] riscv64 native fastdebug debug-level build with gcc14

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341880](https://bugs.openjdk.org/browse/JDK-8341880): RISC-V: riscv_vector.h native build fails with gcc13 after JDK-8320500 (**Bug** - P2)

### Reviewers
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) Review applies to [b9484edd](https://git.openjdk.org/jdk/pull/21442/files/b9484eddea9ed538440411fd712987b2be331015)

### Contributors
 * Fei Yang `<fyang@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21442/head:pull/21442` \
`$ git checkout pull/21442`

Update a local copy of the PR: \
`$ git checkout pull/21442` \
`$ git pull https://git.openjdk.org/jdk.git pull/21442/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21442`

View PR using the GUI difftool: \
`$ git pr show -t 21442`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21442.diff">https://git.openjdk.org/jdk/pull/21442.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21442#issuecomment-2404322811)

### CODE REVIEW

**Comment by sendaoYan** [comment]
> What we can do is to test if rvv intrinsic is really supported as we expect, e.g. if "#include <riscv_vector.h>" is right, if it's true then we can still add `-march=rv64gcv` flag, otherwise do not add the flag. Do you mind to construct such a patch? please kindly let me know. Otherwise I can fix the issue.

I can try to construct that patch. Is there any similar example that test if "#include <something>" works or not.

File: make/modules/jdk.incubator.vector/Lib.gmk [hunk 1]
```diff
@@ -50,7 +50,6 @@ ifeq ($(call isTargetOs, linux)+$(call isTargetCpu, riscv64)+$(INCLUDE_COMPILER2
       EXTRA_SRC := libsleef/generated, \
       DISABLED_WARNINGS_gcc := unused-function sign-compare tautological-compare ignored-qualifiers, \
       DISABLED_WARNINGS_clang := unused-function sign-compare tautological-compare ignored-qualifiers, \
-      CFLAGS := -march=rv64gcv, \
```
Comment (Hamlin-Li) [inline make/modules/jdk.incubator.vector/Lib.gmk:53]: diff --git a/make/autoconf/flags-cflags.m4 b/make/autoconf/flags-cflags.m4
index 992a0282c04..2c54b356c5d 100644
--- a/make/autoconf/flags-cflags.m4
+++ b/make/autoconf/flags-cflags.m4
@@ -912,6 +912,37 @@ AC_DEFUN([FLAGS_SETUP_CFLAGS_CPU_DEP],
         IF_FALSE: [$2FDLIBM_CFLAGS=""])
   fi
   AC_SUBST($2FDLIBM_CFLAGS)
+
+  # Check that the compiler support -march=rv64gcv and riscv_vector.h
+  if test "x${OPENJDK_TARGET_CPU}" = "xriscv64"; then
+    if test "x$TOOLCHAIN_TYPE" = xgcc || test "x$TOOLCHAIN_TYPE" = xclang; then
+      AC_LANG_PUSH(C)
+      OLD_CFLAGS="$CFLAGS"
+      CFLAGS="$CFLAGS -march=rv64gcv"
+      AC_MSG_CHECKING([is rvv intrinsic supported])
+      AC_COMPILE_IFELSE([AC_LANG_PROGRAM([#include <riscv_vector.h>],
+          [
+            const char *src = (const char*) malloc(1024);
+            char *dst = (char*) malloc(1024);
+            size_t vl = __riscv_vsetvl_e8m8(16);
+            vuint8m8_t vec_src = __riscv_vle8_v_u8m8(src, vl);
+            __riscv_vse8_v_u8m8(dst, vec_src, vl);
+            return strlen(dst);
+          ])],
+          [
+            AC_MSG_RESULT([yes])
+            $2RVV_INTRINSICS_CFLAGS="-march=rv64gcv"
+          ],
+          [
+            AC_MSG_RESULT([no])
+            $2RVV_INTRINSICS_CFLAGS=""
+          ]
+      )
+      CFLAGS="$OLD_CFLAGS"
+      AC_LANG_POP(C)
+    fi
+  fi
+  AC_SUBST($2RVV_INTRINSICS_CFLAGS)
 ])
 
 # FLAGS_SETUP_GCC6_COMPILER_FLAGS([PREFIX])
diff --git a/make/autoconf/spec.gmk.template b/make/autoconf/spec.gmk.template
index 20b1d00aa89..a95f5cb5793 100644
--- a/make/autoconf/spec.gmk.template
+++ b/make/autoconf/spec.gmk.template
@@ -827,6 +827,9 @@ OS_VERSION_MAJOR := @OS_VERSION_MAJOR@
 OS_VERSION_MINOR := @OS_VERSION_MINOR@
 OS_VERSION_MICRO := @OS_VERSION_MICRO@
 
+# Riscv: RVV intrinsics flags
+RVV_INTRINSICS_CFLAGS := @RVV_INTRINSICS_CFLAGS@
+
 # Images directory definitions
 JDK_IMAGE_SUBDIR := jdk
 JRE_IMAGE_SUBDIR := jre
diff --git a/make/modules/jdk.incubator.vector/Lib.gmk b/make/modules/jdk.incubator.vector/Lib.gmk
index c6c6103a301..6b689a32a77 100644
--- a/make/modules/jdk.incubator.vector/Lib.gmk
+++ b/make/modules/jdk.incubator.vector/Lib.gmk
@@ -47,6 +47,7 @@ ifeq ($(call isTargetOs, linux)+$(call isTargetCpu, riscv64)+$(INCLUDE_COMPILER2
       DISABLED_WARNINGS_gcc := unused-function sign-compare tautological-compare ignored-qualifiers, \
       DISABLED_WARNINGS_clang := unused-function sign-compare tautological-compare ignored-qualifiers, \
       CFLAGS := $(CFLAGS_JDKLIB) -march=rv64gcv, \
+      vector_math_rvv.c_CFLAGS := $(RVV_INTRINSICS_CFLAGS), \
       LDFLAGS := $(LDFLAGS_JDKLIB) \
           $(call SET_SHARED_LIBRARY_ORIGIN), \
       LIBS := $(JDKLIB_LIBS) \

**Review by Hamlin-Li — COMMENTED** [review]
Can you try something like below?

**Comment by RealFYang** [comment]
FYI: I witnessed some different kinds of compile errors about libsleef while building with GCC version 13.2.0 (Ubuntu 13.2.0-23ubuntu4) shipped with Ubuntu 24.04.1 LTS riscv64. Seems that GCC-13 doesn't have a good support for these RVV intrinsics as it does not recognize quite a few of them. If libsleef claims GCC-14+, maybe it's more reasonable to simply check the GCC version instead requiring 14+ when configuring openjdk? 

```
cc1: all warnings being treated as errors
gmake[3]: *** [Lib.gmk:48: /home/ubuntu/jdk/build/linux-riscv64-server-release/support/native/jdk.incubator.vector/libsleef/vector_math_rvv.o] Error 1
gmake[2]: *** [make/Main.gmk:191: jdk.incubator.vector-libs] Error 2
gmake[2]: *** Waiting for unfinished jobs....

ERROR: Build failed for target 'images' in configuration 'linux-riscv64-server-release' (exit code 2)

=== Output from failing command(s) repeated here ===
* For target support_native_jdk.incubator.vector_libsleef_vector_math_rvv.o:
In file included from /home/ubuntu/jdk/src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c:41:
/home/ubuntu/jdk/src/jdk.incubator.vector/linux/native/libsleef/lib/../generated/sleefinline_rvvm1.h:1146:28: error: unknown type name ‘vfloat64m1x4_t’
 1146 | static SLEEF_ALWAYS_INLINE vfloat64m1x4_t __riscv_vcreate_v_f64m1x4(vfloat64m1_t x, vfloat64m1_t y, vfloat64m1_t z, vfloat64m1_t w) {
      |                            ^~~~~~~~~~~~~~
/home/ubuntu/jdk/src/jdk.incubator.vector/linux/native/libsleef/lib/../generated/sleefinline_rvvm1.h: In function ‘__riscv_vcreate_v_f64m1x4’:
/home/ubuntu/jdk/src/jdk.incubator.vector/linux/native/libsleef/lib/../generated/sleefinline_rvvm1.h:1147:5: error: unknown type name ‘vfloat64m1x4_t’; did you mean ‘vfloat64m4_t’?
 1147 |     vfloat64m1x4_t unused;
      |     ^~~~~~~~~~~~~~
      |     vfloat64m4_t
/home/ubuntu/jdk/src/jdk.incubator.vector/linux/native/libsleef/lib/../generated/sleefinline_rvvm1.h:1148:12: error: implicit declaration of function ‘__riscv_vset’; did you mean ‘__riscv_vlenb’? [-Werror=implicit-function-declaration]
 1148 |     return __riscv_vset(__riscv_vset(__riscv_vset(__riscv_vset(unused, 0, x), 1, y), 2, z), 3, w);
      |            ^~~~~~~~~~~~
      |            __riscv_vlenb
/home/ubuntu/jdk/src/jdk.incubator.vector/linux/native/libsleef/lib/../generated/sleefinline_rvvm1.h: At top level:
/home/ubuntu/jdk/src/jdk.incubator.vector/linux/native/libsleef/lib/../generated/sleefinline_rvvm1.h:1150:28: error: unknown type name ‘vfloat64m2x4_t’
   ... (rest of output omitted)
```

**Comment by Hamlin-Li** [comment]
> FYI: I witnessed some different kinds of compile errors about libsleef while building with GCC version 13.2.0 (Ubuntu 13.2.0-23ubuntu4) shipped with Ubuntu 24.04.1 LTS riscv64. Seems that GCC-13 doesn't have a good support for these RVV intrinsics as it does not recognize quite a few of them. If libsleef claims GCC-14+, maybe it's more reasonable to simply check the GCC version instead requiring 14+ when configuring openjdk?

From https://github.com/shibatch/sleef, it state clearly that gcc 14+ is required, I think we should be fine to check gcc version.

For llvm, there is no version requirement stated in the page, it could still require a little `test` program to tell whether rvv intrinsicv are really supported or not (If `__riscv_v_intrinsic` can not be used as a test macro for that purpose like gcc(e.g. gcc-13) ).

**Comment by RealFYang** [comment]
> > FYI: I witnessed some different kinds of compile errors about libsleef while building with GCC version 13.2.0 (Ubuntu 13.2.0-23ubuntu4) shipped with Ubuntu 24.04.1 LTS riscv64. Seems that GCC-13 doesn't have a good support for these RVV intrinsics as it does not recognize quite a few of them. If libsleef claims GCC-14+, maybe it's more reasonable to simply check the GCC version instead requiring 14+ when configuring openjdk?
> 
> From https://github.com/shibatch/sleef, it state clearly that gcc 14+ is required, I think we should be fine to check gcc version.
> 
> For llvm, there is no version requirement stated in the page, it could still require a little `test` program to tell whether rvv intrinsicv are really supported or not (If `__riscv_v_intrinsic` can not be used as a test macro for that purpose like gcc(e.g. gcc-13) ).

I see they claim LLVM 17+ from the page:
```
Similarly, the RISC-V interface in SLEEF is based on version 1.0 of the intrinsics, which is only supported from llvm version 17 and gcc version 14 onwards.
```

**Comment by Hamlin-Li** [comment]
> FYI: I see they claim LLVM 17+ from the page:

That's great, then it should be fine to just check the compiler version and enable `-march=rv64gcv` conditaionally.

**Comment by RealFYang** [comment]
So I tried following change, seems to work:
```
diff --git a/src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c b/src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c
index 4515457fa89..4911d45dfa4 100644
--- a/src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c
+++ b/src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c
@@ -31,6 +31,8 @@
 // At run-time, if the library is found and the bridge functions are available in the
 // library, then the java vector API will call into the bridge functions and sleef.

+#if __GNUC__ >= 14 || (defined(__clang_major__) && __clang_major__ >= 17)
+
 #ifdef __riscv_v_intrinsic

 #include <stdint.h>
@@ -118,3 +120,5 @@ DEFINE_VECTOR_MATH_BINARY_RVV(hypotdx_u05, vdouble_rvvm1_sleef)
 #undef DEFINE_VECTOR_MATH_BINARY_RVV

 #endif /* __riscv_v_intrinsic */
+
+#endif
```

**Comment by sendaoYan** [comment]
> So I tried following change, seems to work:

This change works for native gcc13 and native gcc14.

Size of vector_math_rvv.o build from gcc14
```
-rw-r--r-- 1 yansendao yansendao 5.0M Oct 11 13:31 ./linux-riscv64-server-release/support/native/jdk.incubator.vector/libsleef/vector_math_rvv.o
```

Size of vector_math_rvv.o build from gcc13
```
-rw-r--r-- 1 yansendao yansendao 1.9K Oct 11 12:12 ./linux-riscv64-server-release/support/native/jdk.incubator.vector/libsleef/vector_math_rvv.o
```

File: make/modules/jdk.incubator.vector/Lib.gmk (hunk 1 above)
Comment (sendaoYan) [reply make/modules/jdk.incubator.vector/Lib.gmk]: Check the gcc and clang version seems will be enough.

**Review by RealFYang — APPROVED** [review]
LGTM. Thanks.

File: src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c [hunk 2]
```diff
@@ -118,3 +120,5 @@ DEFINE_VECTOR_MATH_BINARY_RVV(hypotdx_u05, vdouble_rvvm1_sleef)
 #undef DEFINE_VECTOR_MATH_BINARY_RVV
 
 #endif /* __riscv_v_intrinsic */
+
+#endif
```
Comment (Hamlin-Li) [inline src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c:124]: A tailing blank line is needed?

**Review by Hamlin-Li — COMMENTED** [review]
Thanks!
Looks good, just one minor comment.

File: src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c (hunk 2 above)
Comment (sendaoYan) [reply src/jdk.incubator.vector/linux/native/libsleef/lib/vector_math_rvv.c]: Sorry. The tailing blank line has been added, and add a comment for `#endif`

**Comment by sendaoYan** [comment]
The tailing blank line has been added, and add a comment for the second`#endif`

**Review by Hamlin-Li — APPROVED** [review]
Thanks!

**Comment by sendaoYan** [comment]
Thanks all for the review.

**Pull request closed without merging**